### PR TITLE
[WIP]A decomposition of lr scheduler

### PR DESCRIPTION
**Pull request by hetong007**
Following the data transformation of data loader, I'm decomposing the learning rate scheduler into basic blocks (constant, linear, cosine, poly, and possibly more).

To construct a learning rate scheduler with multiple stages, one can `Compose` blocks together. This offers flexibility in the learning rate schedule, for instance now we can design a learning rate schedule as:
1. warmup for 5 epochs
2. cosine schedule for 100 epochs
3. constant schedule for 5 epochs
4. smaller constant schedule for 5 epochs

TODO:
- [x] Test with image classification training
- [x] Modify affected detection/segmentation scripts
- [x] Unit tests.

### CODE REVIEW

**Comment by mli** [comment]
Job PR-353-11 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-353/11/index.html 
Code coverage of this PR:  vs. Master:

**Comment by zhanghang1989** [comment]
could you also update the code in semantic segmentation training
https://github.com/dmlc/gluon-cv/blob/master/scripts/segmentation/train.py#L149

**Comment by hetong007** [comment]
I'll do a full scan on scripts and update accordingly.

File: gluoncv/utils/lr_scheduler.py [hunk 1]
```diff
@@ -5,104 +5,114 @@
...
-                 warmup_epochs=0, warmup_lr=0, warmup_mode='linear'):
+    def __init__(self, mode, baselr, targetlr=0, niters=0, offset=0, power=2,
+                 step=None, step_factor=0.1):
         super(LRScheduler, self).__init__()
-        assert(mode in ['step', 'poly', 'cosine'])
-        assert(warmup_mode in ['linear', 'constant'])
+        assert(mode in ['constant', 'step', 'linear', 'poly', 'cosine'])
 
         self.mode = mode
         self.baselr = baselr
```
Comment (eric-haibin-lin) [inline gluoncv/utils/lr_scheduler.py]: the parent class uses `self.base_lr` instead of `self.baselr`
https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/lr_scheduler.py#L43

Comment (hetong007) [reply gluoncv/utils/lr_scheduler.py]: Will match the parent class and warn users for using `self.baselr`.

**Comment by mli** [comment]
Job PR-353-20 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-353/20/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-353-25 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-353/25/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-353-27 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-353/27/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-353-28 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-353/28/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-353-29 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-353/29/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-353-30 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-353/30/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-353-31 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-353/31/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-353-32 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-353/32/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-353-45 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-353/45/index.html 
Code coverage of this PR:  vs. Master:

**Comment by hetong007** [comment]
@zhanghang1989 @eric-haibin-lin @zhreshold I'm about to merge this PR for 0.4. Please review and comment if you have suggestions.

**Comment by mli** [comment]
Job PR-353-46 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-353/46/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-353-47 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-353/47/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-353-43 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-353/43/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-353-49 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-353/49/index.html 
Code coverage of this PR:  vs. Master:
